crates/hir-ty/src/infer/expr.rs

```diff
@@ -564,9 +564,29 @@ impl<'a> InferenceContext<'a> {
                 let inner_ty = self.infer_expr_inner(*expr, &Expectation::none());
                 self.resolve_associated_type(inner_ty, self.resolve_future_future_output())
             }
-            Expr::Try { expr } => {
-                let inner_ty = self.infer_expr_inner(*expr, &Expectation::none());
-                self.resolve_associated_type(inner_ty, self.resolve_ops_try_ok())
+            &Expr::Try { expr } => {
+                let inner_ty = self.infer_expr_inner(expr, &Expectation::none());
+                match self.resolve_try_impl_for(inner_ty.clone()) {
+                    Some((_, Some((output, residual)))) => {
+                        if let Some((_trait, false)) =
+                            self.implements_from_residual(self.return_ty.clone(), residual)
```

Comment:
it might not be possible to solve this uniquely at this point, which means we'll miss some cases here -- the proper place to emit these diagnostics would be in `resolve_obligations_as_possible`, which of course means we have to track where the obligations come from

Reply:
Hmm, ye I see what you mean. The way this is done shouldn't cause wrong diagnostics though if I see this right, only miss some?

Reply:
Yes